### PR TITLE
[WIP] Refactoring Search API to accept query parameters through a map.

### PR DESCRIPTION
**Pull request by dhruvsinghal**
@pengwynn Here's the barebones modification to the search API, just to check if I'm on the right track.

### CODE REVIEW

File: octokit/search.go [hunk 1]
```diff
@@ -5,44 +5,44 @@ import (
 	"net/url"
 )
 
-var SearchURL = Hyperlink("search{/type}?q={query}{&page,per_page,sort,order}")
+var SearchURL = Hyperlink("search{/type}?q={query}&page={page}&per_page={per_page}&sort={sort}&order={order}")
```
Comment (pengwynn) [inline octokit/search.go]: No need to change these. That's the correct format for URI template params like this. The expansion will handle the `?` and `&` automatically.

File: octokit/search.go [hunk 2]
```diff
@@ -5,44 +5,44 @@ import (
...
 
-func (c *Client) Search(url *url.URL) (searches *SearchService) {
-	searches = &SearchService{client: c, URL: url}
+func (c *Client) Search(uriTemplate *Hyperlink) (searches *SearchService) {
```
Comment (pengwynn) [inline octokit/search.go]: Let's explore moving the hyperlink to the actual method call. Something like:

```go

func (g *SearchService) Users(h Hyperlink, m M) (userSearchResults UserSearchResults, result *Result) {
    if h == nil {
        h = SearchURL
    }
    url, err := h.Expand(m)
    ...
```

**Comment by dhruvsinghal** [comment]
Please take a look. Note that only the user search has been implemented.

**Comment by dhruvsinghal** [comment]
Please take a look. I've refactored the code as we discussed.

**Comment by dhruvsinghal** [comment]
@pengwynn Please take a look. I've refactored the code as we discussed.

File: octokit/search.go [hunk 3]
```diff
@@ -2,47 +2,60 @@ package octokit
 
 import (
 	"github.com/jingweno/go-sawyer/hypermedia"
-	"net/url"
 )
 
-var SearchURL = Hyperlink("search{/type}?q={query}{&page,per_page,sort,order}")
+var SearchURITemplate = "search{/type}?q={query}{&page,per_page,sort,order}"
```
Comment (pengwynn) [inline octokit/search.go]: I like keeping this a [`Hyperlink`][link] for convenience.

[link]: https://github.com/octokit/go-octokit/blob/master/octokit/hyperlink.go

File: octokit/search.go [hunk 4]
```diff
@@ -2,47 +2,60 @@ package octokit
...
-	return
+func (c *Client) Search(uriTemplate string) *SearchService {
+	return &SearchService{&GenericService{client: c, uriTemplate: uriTemplate}}
 }
 
 // A service to return search records
 type SearchService struct {
-	client *Client
-	URL    *url.URL
+	*GenericService
```
Comment (pengwynn) [inline octokit/search.go]: We _may_ end up here, but I'd like to see a couple of implementations before we extract this. @lalalah's work on Gists might help.

**Pull request closed without merging**